mteb/encoder_interface.py

```diff
@@ -0,0 +1,60 @@
+from __future__ import annotations
+
+from typing import Any, Protocol, runtime_checkable
+
+import numpy as np
+import torch
+
+
+@runtime_checkable
+class Encoder(Protocol):
+    """The interface for an encoder in MTEB."""
+
+    def encode(
+        self, sentences: list[str], prompt: str, **kwargs: Any
```

Comment:
prompt does not really allow the model to generate a custom prompt for their task.

A solution might be to add an optional method `.generate_prompt(task_name: str) -> prompt`?

Reply:
Like with #457, it would be great if the prompt/instruction could also be `List[str]` to account for when the data provides it

Reply:
Should we create an issue on SentenceTransformers on this?

Reply:
Another option to supply the prompt would also be to do something like:

```
bench = MTEB(..., prompts={"task1": "my prompt for task1", "task2": ...})
```

Reply:
Actually, it seems like SentenceTransformer encode has a "prompt_name" as input (thanks to @isaac-chung for pointing that out), which takes a string and looks up the corresponding prompt in a dictionary. I would just use this instead.

Reply:
The example above still is only per-task (dataset) instead of per-instance.  It also looks like it forces you to only submit one `prompt_name` that selects a given string prompt to use for the entire dataset. This doesn't allow each query in one dataset to come with its own different instruction per query.

If I'm misunderstanding and the API allows you to put {query->prompt} dict for each query in the dataset and it loops over each one then great! Otherwise I think it may have to be a PR to SentenceTransformers.

Is this a correct assessment of SentenceTransformers @isaac-chung?

Reply:
Yea this would just be a a singular prompt for the task  -which I believe is valid for all tasks except for the FollowIR?